Distmaker: add CiviCRM Standalone support

mlutfy opened the pull request:
Overview
----------------------------------------

Adds support for building CiviCRM Standalone tar.gz archives for nightly/rc/stable releases.

Before
----------------------------------------

No CiviCRM Standalone builds.

After
----------------------------------------

`civicrm-standalone-5.66-standalone.tar.gz` :sunrise_over_mountains: 

Technical Details
----------------------------------------

I wasn't too sure what directory structure to follow.

Based on [civicrm standalone installation docs](https://docs.civicrm.org/installation/en/latest/standalone/), I figured it would be best to not use the composer template, and use the more minimal method. It still supports hosting civicrm and private files outside the webroot.

Comments
-------------------

This may be of a surprise, but I did test locally:

```
./src/distmaker/distmaker.sh standalone
```

and then I checked the resulting tar, but I did not yet test using the tar to install CiviCRM from scratch.

mlutfy committed:
Distmaker: add CiviCRM Standalone support

seamuslee001 commented on distmaker/dists/standalone.sh:
```diff
@@ -0,0 +1,32 @@
+dm_install_packages "$SRC/packages" "$TRG/civicrm/core/packages"
+dm_install_vendor "$SRC/vendor" "$TRG/civicrm/core/vendor"
+dm_install_bower "$SRC/bower_components" "$TRG/civicrm/core/bower_components"
+dm_install_cvext com.iatspayments.civicrm "$TRG/civicrm/core/ext/iatspayments"
```
Do we need to include the standalone extension?

mlutfy replied:
@seamuslee001 it's part of core: https://github.com/civicrm/civicrm-core/tree/master/ext/standaloneusers

totten replied:
In this file, it has:

```bash
dm_install_coreext "$SRC" "$TRG/civicrm/core" $(dm_core_exts)
```

which reads `distmaker/core-ext.txt` -- and yes, `standaloneusers` is listed there too. ✅ 

(Fun aside: I think `standaloneusers` was added to `core-ext.txt` to ensure that `set-version.php` maintains it. However, this does imply that *all* tarballs include it. That's a bit extraneous. But as long the ext is hidden, it doesn't feel like a real problem.)

colemanw replied:
@totten so should this be merged?

totten replied:
@colemanw Not yet. (a) I don't think anyon has `r-run` the output. (b) it probably needs [some change to serve static assets](https://github.com/civicrm/civicrm-core/pull/27104#issuecomment-1687326121) -- e.g. use the `web/core/` folder or revise `htaccess`.

totten replied:
(But this subthread can be resolved. I ran distmaker and confirmed that `ext/standaloneusers`  is included.)

totten commented:
> Based on [civicrm standalone installation docs](https://docs.civicrm.org/installation/en/latest/standalone/), I figured it would be best to not use the composer template, and use the more minimal method. It still supports hosting civicrm and private files outside the webroot....

Yeah, I can see the rationale for skipping the composer template on this build. (*There's kind of a chicken/egg problem with having distmaker do a release that depends on composer to fetch the release that distmaker is building...*)

FWIW,  in the layout defined by "Download CiviCRM directly (Standard web-server)", it actually does store  civicrm in the public web folder (i.e. `web/core` and `web/core/packages`). A working system needs some way to serve static assets (CSS, JS, PNG, etc). Major options are:

1. Put`civicrm` in the public folder.
    * __Example__: Download CiviCRM directly (Standard web-server)
2. Put `civicrm` in a private folder and define HTTPD rules for static assets.
    * __Example__: Download CiviCRM directly (PHP built-in web-server)
3. Put `civicrm` in a private folder and define sync rules for static assets.
    *  __Example__:  "Download the starter template" with  `civicrm-asset-plugin`

IMHO, it's OK to do the "easy thing" and use `web/core/` and `web/core/packages/`. A huge portion of Civi sites (D7/BD/WP/J) do use a public folder, and (IIRC) it's been a number of years since we actually had an issue stemming from that. It's not ideal, but it's not terrible.

If you do want a way to use a private folder without the composer plugin, then... you could try to sort out `.htaccess` and nginx options. I'm not sure how easy/hard that would be. Here's an [example configuration from PHP 's cli-server](https://github.com/civicrm/civicrm-core/blob/79b6d5084142b664d86b927e396772085c5e073d/tools/standalone/router.php#L54-L64). It could be locked down a bit further. (_Allow folders `js/` `ang/` `ext/`; disallow folders `CRM/` `Civi/`. Allow extensions `*.js` `*.css`; disallow extensions `*.php` `*.sh` `*.conf`._)

totten commented:
> the installer looks for files such as `/assets/civicrm/core/setup/res/updated-logo.jpg` - I noticed that `setup/plugins/init/Standalone.civi-setup.php` sets the path, but it's not detecting `civicrm/civicrm-asset-plugin` and going in the "else", so I'm not sure where it's being set.

I think the problem is higher up -- when `index.php` calls the setup APIs to load the GUI, it [sets `/assets/civicrm/core`](https://github.com/civicrm/civicrm-core/blob/ccabb1d6ef5b3f3f11a104197b8458c29b8afa2e/setup/res/index.php.txt#L121-L132). 

Maybe there's a way for GUI loader to use the output from `Standalone.civi-setup.php`? e.g.

```php
// $coreUrl = '/assets/civicrm/core';

\Civi\Setup::init([ ... ]); // Fires `civi.setup.init` and thus `Standalone.civi-setup.php`

$coreUrl = \Civi\Setup::instance()->getModel()->mandatorySettings['userFrameworkResourceURL'];
```

mlutfy commented:
@totten I think we're going wee down the rabbit hole on the warnings. The actual distmaker code works, and the installer is usable. The rest are improvements we can do along the way?

mlutfy commented:
@totten Ignore my last comment, sorry I need more coffee ;)

- Fixed the assets bug based on your example (I didn't test completely, but tested that the tar.gz output works)
- Removed the `cmsDb` warning for Standalone

I would circle back for the file permission warnings in a separate PR. There may be other issues, but I would prefer treating them separately.

One thing that seems to be missing is the web-accessible `web/upload` directory (which must be chowned/chmodded too), and oddly the installer did not complain about it.

totten commented on setup/plugins/checkRequirements/CheckDbWellFormed.civi-setup.php:
```diff
@@ -13,7 +13,12 @@
-    $dbFields = array('db', 'cmsDb');
+    $dbFields = ['db'];
+
+    if ($e->getModel()->cms !== 'Standalone') {
```
Test failed because `CIVICRM_UF` was not set. The `$model->cms` is the pre-installation equivalent. Pushed a tweak. for that.

mlutfy replied:
@totten thanks :)

totten commented:
Cool. I did a full test-run (build `civicrm-5.66.alpha1-standalone.tar.gz` on one project; and use that to spin-up another local project on Apache VDR). The site was installed and active. 😃

mlutfy commented:
@totten thanks! Now that nightly archives are working (and on gcloud), I added it to download.civicrm.org: https://github.com/civicrm/civicrm-dist-manager/pull/12
